runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java

```diff
@@ -93,6 +93,8 @@
         // Next, we sync the object, which will bring the object
         // to the correct version, reflecting any optimistic
         // updates.
+        // Get snapshot timestamp in advance so it is not performed under the VLO lock
+        long ts = getSnapshotTimestamp();
```

Comment:
hmm, can't you just remove the lazy evaluation here
https://github.com/CorfuDB/CorfuDB/blob/master/runtime/src/main/java/org/corfudb/runtime/object/transactions/AbstractTransactionalContext.java#L105 ?

Reply:
If we remove the lazy getter we would be having a second RPC call in the access function whenever an update function is applied, but still if you see my previous comment I'm wondering if it ain't the actual right thing to do in order to sync the object..

Reply:
If you removed the lazy evaluation, the value would no longer be caches, and therefore result in a rpc on _every_ object access

Reply:
Cached* not to mention it would break correctness

Reply:
I don't see how this is related to caching, lombok's lazy eval dictates when a class field is initialized (i.e. construction vs first access time initialization). If you delombok the code you'll see this:

private final long snapshotTimestamp = obtainSnapshotTimestamp();

public long getSnapshotTimestamp() {
        return this.snapshotTimestamp;
} 

In other words, the transaction snapshot would be evaluate once the class is created and then subsequent calls to getSnapshotTimestamp wouldn't actually result in an rpc, which is what we wanted to do, or am I missing something?

Reply:
I see your point, you say moving it out of the lock is equivalent to removing the lazy evaluation and removing the lombok getter annotation. I agree, however, wouldn't this increase our chances of aborted transactions? as it might be computed way in advance to the access (according to any client business logic running within a TX). 

On the other hand, either solution we adopt which both attempt to do a single RPC, would you agree that the syncWithRetryUnsafe is not really doing anything? we are passing the same snapshotTimestamp, so no way to sync back or forward.

Reply:
Yes, I think they are equivalent and "cleaner". True, that can happen, but I think its also fine to assume transactional semantics right after tx.begin is called as opposed to first read/write. 

Why not? Notice that there are two time snapshots, one for the tx and the other that belongs to the VLO's stream. Consider the case where you have a stream at version 10 and you have a transaction that starts at version 5, then the direct access will evaluate to false since o.getVersionUnsafe() != streamReadPosition , then the update function is going to be applied, which will call syncWithRetryUnsafe(o (stream VLO ts 10) , 5 (ts of tx), .....) and that will call syncObjectUnsafe which will keep rolling back the smr stream from 10 (vlo stream version) to 5 (tx ts). Does this seem right or did i misunderstand your point?

Reply:
Regarding the first point,I'm fine with this if we all agree this is the right semantic for an Optimistic Tx, I'll submit a PR with this change.

Thanks for the explanation, this example does give me a use case in which I see it.

The scenario I was thinking of was this. Let's assume you have a TX that starts at V5 and the stream is also at v5, since the direct access function is evaluated within the lock, suppose we are not able to return because we cannot validate the lock (optimistic read), the only reason why we wouldn't be able to do so is because the object was modified (by this client), so upgrading to a writelock and failing to do the direct access again would lead to syncWithRetryUnsafe which I thought would attempt to bring the object to the latest version (let's say v6 as it was updated by another thread and we are aware of it). But know that I see your example, I imagine in this case we just rollback to v5 and eventually this TX will be aborted...

Reply:
The reason we obtain tx snapshot lazily for R-W transactions is that a write-only r-w transaction  doesn't need a snapshot, getting a snapshot would be a wasted RPC to the sequencer. We should leave as is.

Reply:
I was just responding to an earlier comment before the merge, I think its fine for now.

For your example, let's say T1 tries to read V5 and T2 move the object to V6, then:
T1: start optimistic read
T2: acquire write lock and apply the update and relinquish the lock.
T1: apply direct access function and fail the lock validation. 
T1: acquire the write lock and apply the update function (i.e. syncWithRetryUnsafe -> rollbackStreamUnsafe)
T1: apply the access function (after lock upgrade)
 
Notice after the failed validation (with direct access function), it will apply the update function which synchronizes the versions (i.e. roll back from V6 to V5). I think the key point here is that optimistic reading path never changes the objects state/version.

Reply:
@no2chem I see, but that's not happening in the current code base, since transactions need TxResolutionInfo, which depend on getSnapshotTimestamp.